doc/man/emacs.1
.\" See section COPYING for copyright and redistribution information.
.TH EMACS 1 "2007 April 13" "GNU Emacs 24.1"
.
.
.SH NAME
emacs \- GNU project Emacs
.
.
.SH SYNOPSIS
.B emacs
[
.I command-line switches
] [
.I files ...\&
]
.
.
.SH DESCRIPTION
.I GNU Emacs
is a version of
.IR Emacs ,
written by the author of the original (PDP-10)
.IR Emacs ,
Richard Stallman.
The user functionality of GNU Emacs encompasses everything other
editors do, and it is easily extensible since its editing commands are
written in Lisp.
.PP
The primary documentation of GNU Emacs is in the GNU Emacs Manual,
which you can read using Info, either from Emacs or as a standalone
program.
Please look there for complete and up-to-date documentation.
This man page is updated only when someone volunteers to do so.
.PP
.I Emacs
has an extensive interactive help facility,
but the facility assumes that you know how to manipulate
.I Emacs
windows and buffers.
CTRL-h or F1 enters the Help facility.
Help Tutorial (CTRL-h t) starts an interactive tutorial to quickly
teach beginners the fundamentals of
.I Emacs.
Help Apropos (CTRL-h a) helps you find a command with a name matching
a given pattern, Help Key (CTRL-h k) describes a given key sequence,
and Help Function (CTRL-h f) describes a given Lisp function.
.PP
.IR "GNU Emacs" 's
many special packages handle mail reading (RMail) and sending (Mail),
outline editing (Outline), compiling (Compile), running subshells
within
.I Emacs
windows (Shell), running a Lisp read-eval-print loop
(Lisp-Interaction-Mode), automated psychotherapy (Doctor), and much more.
.
.SS Emacs Options
The following options are of general interest:
.RS
.TP 8
.I file
Edit
.IR file .
.TP
.BI \-\-file " file\fR,\fP " \-\-find-file " file\fR,\fP " \-\-visit " file"
The same as specifying
.I file
directly as an argument.
.TP
.BI + number
Go to the line specified by
.I number
(do not insert a space between the "+" sign and
the number).
This applies only to the next file specified.
.TP
.BI + line:column
Go to the specified
.I line
and
.IR column .
.TP
.BR \-q ", " \-\-no\-init\-file
Do not load an init file.
.TP
.B \-\-no\-site\-file
Do not load the site-wide startup file.
.TP
.B \-\-no\-desktop
Do not load a saved desktop.
.TP
.BR \-Q ", " \-\-quick
Similar to "\-q \-\-no\-site\-file \-\-no\-splash".  Also, avoid
processing X resources.
.TP
.B \-\-no\-splash
Do not display a splash screen during start-up.
.TP
.B \-\-debug\-init
Enable
.I Emacs
Lisp debugger during the processing of the user init file
.BR ~/.emacs .
This is useful for debugging problems in the init file.
.TP
.BI \-u " user\fR,\fP " \-\-user " user"
Load
.IR user 's
init file.
.TP
.BI \-t " file\fR,\fP " \-\-terminal " file"
Use specified
.I file
as the terminal instead of using stdin/stdout.
This must be the first argument specified in the command line.
.TP
.B \-\-daemon
Start Emacs as a daemon, enabling the Emacs server and disconnecting
from the terminal.  You can then use the emacsclient command to
connect to the server (see
.BR emacsclient (1)).
.TP
.B \-\-version
Display
.I Emacs
version information and exit.
.TP
.B \-\-help
Display this help and exit.
.RE
.PP
The following options are Lisp-oriented
(these options are processed in the order encountered):
.RS
.TP 8
.BI \-f " function\fR,\fP " \-\-funcall " function"
Execute the lisp function
.IR function .
.TP
.BI \-l " file\fR,\fP " \-\-load " file"
Load the lisp code in the file
.IR file .
.TP
.BI \-\-eval " expr\fR,\fP " \-\-execute " expr"
Evaluate the Lisp expression
.IR expr .
.RE
.PP
The following options are useful when running
.I Emacs
as a batch editor:
.RS
.TP 8
.B \-\-batch
Edit in batch mode.
The editor will send messages to stderr.
You must use \-l and \-f options to specify files to execute
and functions to call.
.TP
.BI \-\-script " file"
Run
.I file
as an Emacs Lisp script.
.TP
.BI \-\-insert " file"
Insert contents of
.I file
into the current buffer.
.TP
.B \-\-kill
Exit
.I Emacs
while in batch mode.
.TP
.BI \-L " dir\fR,\fP " \-\-directory " dir"
Add
.I dir
to the list of directories
.I Emacs
searches for Lisp files.
.RE
.
.\" START DELETING HERE IF YOU'RE NOT USING X
.SS Using Emacs with X
.I Emacs
has been tailored to work well with the X window system.
If you run
.I Emacs
from under X windows, it will create its own X window to
display in.
You will probably want to start the editor as a background
process so that you can continue using your original window.
.PP
.I Emacs
can be started with the following X switches:
.RS
.TP 8
.BI \-\-name " name"
Specify the name which should be assigned to the initial
.I Emacs
window.
This controls looking up X resources as well as the window title.
.TP
.BI \-T " name\fR,\fP " \-\-title " name"
Specify the title for the initial X window.
.TP
.BR \-r ", " \-rv ", " \-\-reverse\-video
Display the
.I Emacs
window in reverse video.
.TP
.BI \-fn " font\fR,\fP " \-\-font " font"
Set the
.I Emacs
window's font to that specified by
.IR font .
You will find the various
.I X
fonts in the
.I /usr/lib/X11/fonts
directory.
Note that
.I Emacs
will only accept fixed width fonts.
Under the X11 Release 4 font-naming conventions, any font with the
value "m" or "c" in the eleventh field of the font name is a fixed
width font.
Furthermore, fonts whose name are of the form
.IR width x height
are generally fixed width, as is the font
.IR fixed .
See
.BR xlsfonts (1)
for more information.

When you specify a font, be sure to put a space between the
switch and the font name.
.TP
.BI \-\-xrm " resources"
Set additional X resources.
.TP
.BI "\-\-color\fR,\fP \-\-color=" mode
Override color mode for character terminals;
.I mode
defaults to `auto', and can also be `never', `auto', `always',
or a mode name like `ansi8'.
.TP
.BI \-bw " pixels\fR,\fP " \-\-border\-width " pixels"
Set the
.I Emacs
window's border width to the number of pixels specified by
.IR pixels .
Defaults to one pixel on each side of the window.
.TP
.BI \-ib " pixels\fR,\fP " \-\-internal\-border " pixels"
Set the window's internal border width to the number of pixels specified
by
.IR pixels .
Defaults to one pixel of padding on each side of the window.
.TP
.BI \-g " geometry\fR,\fP " \-\-geometry " geometry"
Set the
.I Emacs
window's width, height, and position as specified.
The geometry specification is in the standard X format; see
.BR X (7)
for more information.
The width and height are specified in characters; the default is
80 by 24.
See the Emacs manual, section "Options for Window Size and Position",
for information on how window sizes interact
with selecting or deselecting the tool bar and menu bar.
.TP
.BI \-lsp " pixels\fR,\fP " \-\-line\-spacing " pixels"
Additional space to put between lines.
.TP
.BR \-vb ", " \-\-vertical\-scroll\-bars
Enable vertical scrollbars.
.TP
.BR \-fh ", " \-\-fullheight
Make the first frame as high as the screen.
.TP
.BR \-fs ", " \-\-fullscreen
Make the first frame fullscreen.
.TP
.BR \-fw ", " \-\-fullwidth
Make the first frame as wide as the screen.
.TP
.BR \-mm ", " \-\-maximized
Maximize the first frame, like "\-fw \-fh".
.TP
.BI \-fg " color\fR,\fP " \-\-foreground\-color " color"
On color displays, set the color of the text.

Use the command
.I M\-x list\-colors\-display
for a list of valid color names.
.TP
.BI \-bg " color\fR,\fP " \-\-background\-color " color"
On color displays, set the color of the window's background.
.TP
.BI \-bd " color\fR,\fP " \-\-border\-color " color"
On color displays, set the color of the window's border.
.TP
.BI \-cr " color\fR,\fP " \-\-cursor\-color " color"
On color displays, set the color of the window's text cursor.
.TP
.BI \-ms " color\fR,\fP " \-\-mouse\-color " color"
On color displays, set the color of the window's mouse cursor.
.TP
.BI \-d " displayname\fR,\fP " \-\-display " displayname"
Create the
.I Emacs
window on the display specified by
.IR displayname .
Must be the first option specified in the command line.
.TP
.BR \-nbi ", " \-\-no\-bitmap\-icon
Do not use picture of gnu for Emacs icon.
.TP
.B \-\-iconic
Start
.I Emacs
in iconified state.
.TP
.BR \-nbc ", " \-\-no\-blinking\-cursor
Disable blinking cursor.
.TP
.BR \-nw ", " \-\-no\-window\-system
Tell
.I Emacs
not to create a graphical frame.
If you use this switch when invoking
.I Emacs
from an
.BR xterm (1)
window, display is done in that window.
.TP
.BR \-D ", " \-\-basic\-display
This option disables many display features; use it for
debugging Emacs.
.RE
.PP
You can set
.I X
default values for your
.I Emacs
windows in your
.I \.Xresources
file (see
.BR xrdb (1)).
Use the following format:
.IP
.RI emacs. keyword : value
.PP
where
.I value
specifies the default value of
.IR keyword .
.I Emacs
lets you set default values for the following keywords:
.RS
.TP 8
.BR background " (class " Background )
For color displays,
sets the window's background color.
.TP
.BR bitmapIcon " (class " BitmapIcon )
If
.BR bitmapIcon 's
value is set to
.IR on ,
the window will iconify into the "kitchen sink."
.TP
.BR borderColor " (class " BorderColor )
For color displays,
sets the color of the window's border.
.TP
.BR borderWidth " (class " BorderWidth )
Sets the window's border width in pixels.
.TP
.BR cursorColor " (class " Foreground )
For color displays,
sets the color of the window's text cursor.
.TP
.BR cursorBlink " (class " CursorBlink )
Specifies whether to make the cursor blink.
The default is
.IR on .
Use
.I off
or
.I false
to turn cursor blinking off.
.TP
.BR font " (class " Font )
Sets the window's text font.
.TP
.BR foreground " (class " Foreground )
For color displays,
sets the window's text color.
.TP
.BR fullscreen " (class " Fullscreen )
The desired fullscreen size.
The value can be one of
.IR fullboth ,
.IR maximized ,
.IR fullwidth ,
or
.IR fullheight ,
which correspond to the command-line options `\-fs', `\-mm', `\-fw',
and `\-fh', respectively.
Note that this applies to the initial frame only.
.TP
.BR geometry " (class " Geometry )
Sets the geometry of the
.I Emacs
window (as described above).
.TP
.BR iconName " (class " Title )
Sets the icon name for the
.I Emacs
window icon.
.TP
.BR internalBorder " (class " BorderWidth )
Sets the window's internal border width in pixels.
.TP
.BR lineSpacing " (class " LineSpacing )
Additional space ("leading") between lines, in pixels.
.TP
.BR menuBar " (class " MenuBar )
Gives frames menu bars if
.IR on ;
don't have menu bars if
.IR off .
See the Emacs manual, sections "Lucid Resources" and "LessTif
Resources", for how to control the appearance of the menu bar
if you have one.
.TP
.BR minibuffer " (class " Minibuffer )
If
.IR none ,
don't make a minibuffer in this frame.
It will use a separate minibuffer frame instead.
.TP
.BR paneFont " (class " Font )
Font name for menu pane titles, in non-toolkit versions of
.IR Emacs .
.TP
.BR pointerColor " (class " Foreground )
For color displays,
sets the color of the window's mouse cursor.
.TP
.BR privateColormap " (class " PrivateColormap )
If
.IR on ,
use a private color map, in the case where the "default
visual" of class
.B PseudoColor
and
.B Emacs
is using it.
.TP
.BR reverseVideo " (class " ReverseVideo )
If
.BR reverseVideo 's
value is set to
.IR on ,
the window will be displayed in reverse video.
.TP
.BR screenGamma " (class "ScreenGamma )
Gamma correction for colors, equivalent to the frame parameter
`screen\-gamma'.
.TP
.BR scrollBarWidth " (class "ScrollBarWidth )
The scroll bar width in pixels, equivalent to the frame parameter
`scroll\-bar\-width'.
.TP
.BR selectionFont " (class " SelectionFont )
Font name for pop-up menu items, in non-toolkit versions of
.IR Emacs .
(For toolkit versions, see the Emacs manual, sections
"Lucid Resources" and "LessTif Resources".)
.TP
.BR selectionTimeout " (class " SelectionTimeout )
Number of milliseconds to wait for a selection reply.
A value of 0 means wait as long as necessary.
.TP
.BR synchronous " (class " Synchronous )
Run Emacs in synchronous mode if
.IR on .
Synchronous mode is useful for debugging X problems.
.TP
.BR title " (class " Title )
Sets the title of the
.I Emacs
window.
.TP
.BR toolBar " (class " ToolBar )
Number of lines to reserve for the tool bar.
.TP
.BR useXIM " (class " UseXIM )
Turns off use of X input methods (XIM) if
.I false
or
.IR off .
.TP
.BR verticalScrollBars " (class " ScrollBars )
Gives frames scroll bars if
.IR on ;
suppresses scroll bars if
.IR off .
.TP
.BR visualClass " (class " VisualClass )
Specify the "visual" that X should use.
This tells X how to handle colors.
The value should start with one of
.IR TrueColor ,
.IR PseudoColor ,
.IR DirectColor ,
.IR StaticColor ,
.IR GrayScale ,
and
.IR StaticGray ,
followed by
.BI \- depth\fR,\fP
where
.I depth
is the number of color planes.
.RE
.
.SH MANUALS
You can order printed copies of the GNU Emacs Manual from the Free
Software Foundation, which develops GNU software.
See the online store at <http://shop.fsf.org/>.
.br
Your local administrator might also have copies available.
As with all software and publications from FSF, everyone is permitted
to make and distribute copies of the Emacs manual.
The Texinfo source to the manual is also included in the Emacs source
distribution.
.
.
.SH FILES
/usr/local/share/info \(em files for the Info documentation browser.
The complete text of the Emacs reference manual is included in a
convenient tree structured form.
Also includes the Emacs Lisp Reference Manual, useful to anyone
wishing to write programs in the Emacs Lisp extension language,
and the Introduction to Programming in Emacs Lisp.

/usr/local/share/emacs/$VERSION/lisp \(em Lisp source files and
compiled files that define most editing commands.
Some are preloaded; others are autoloaded from this directory when
used.

/usr/local/libexec/emacs/$VERSION/$ARCH \(em various programs that are
used with GNU Emacs.

/usr/local/share/emacs/$VERSION/etc \(em various files of information.

/usr/local/share/emacs/$VERSION/etc/DOC.* \(em contains the documentation
strings for the Lisp primitives and preloaded Lisp functions
of GNU Emacs.
They are stored here to reduce the size of Emacs proper.

/usr/local/share/emacs/$VERSION/etc/SERVICE lists people offering
various services to assist users of GNU Emacs, including education,
troubleshooting, porting and customization.
.
.
.SH BUGS
There is a mailing list, bug-gnu-emacs@gnu.org, for reporting Emacs
bugs and fixes.
But before reporting something as a bug, please try to be sure that
it really is a bug, not a misunderstanding or a deliberate feature.
We ask you to read the section ``Reporting Bugs'' in the Emacs manual
for hints on how and when to report bugs.
Also, include the version number of the Emacs you are running in
\fIevery\fR bug report that you send in.
Bugs tend actually to be fixed if they can be isolated, so it is
in your interest to report them in such a way that they can be
easily reproduced.

Do not expect a personal answer to a bug report.
The purpose of reporting bugs is to get them fixed for everyone
in the next release, if possible.
For personal assistance, look in the SERVICE file (see above) for
a list of people who offer it.

Please do not send anything but bug reports to this mailing list.
For more information about Emacs mailing lists, see the
file /usr/local/share/emacs/$VERSION/etc/MAILINGLISTS.
.
.
.SH UNRESTRICTIONS
.I Emacs
is free; anyone may redistribute copies of
.I Emacs
to anyone under the terms stated in the GNU General Public License,
a copy of which accompanies each copy of
.I Emacs
and which also
appears in the reference manual.
.PP
Copies of
.I Emacs
may sometimes be received packaged with distributions of Unix systems,
but it is never included in the scope of any license covering those
systems.
Such inclusion violates the terms on which distribution is permitted.
In fact, the primary purpose of the GNU General Public License is to
prohibit anyone from attaching any other restrictions to
redistribution of
.IR Emacs .
.PP
Richard Stallman encourages you to improve and extend
.IR Emacs ,
and urges that
you contribute your extensions to the GNU library.
Eventually GNU (Gnu's Not Unix) will be a complete replacement
for Unix.
Everyone will be free to use, copy, study and change the GNU system.
.
.
.SH SEE ALSO
.BR emacsclient (1),
.BR etags (1),
.BR X (7),
.BR xlsfonts (1),
.BR xterm (1),
.BR xrdb (1)
.
.
.SH AUTHORS
.I Emacs
was written by Richard Stallman and the Free Software Foundation.
For detailed credits and acknowledgements, see the GNU Emacs manual.
.
.
.
.SH COPYING
Copyright
.if t \(co
.if n (C)
1995, 1999-2012 Free Software Foundation, Inc.
.PP
Permission is granted to make and distribute verbatim copies of this
document provided the copyright notice and this permission notice are
preserved on all copies.
.PP
Permission is granted to copy and distribute modified versions of
this document under the conditions for verbatim copying, provided that
the entire resulting derived work is distributed under the terms of
a permission notice identical to this one.
.PP
Permission is granted to copy and distribute translations of this
document into another language, under the above conditions for
modified versions, except that this permission notice may be stated
in a translation approved by the Free Software Foundation.
.


etc/NEWS
GNU Emacs NEWS -- history of user-visible changes.

Copyright (C) 2010-2012  Free Software Foundation, Inc.
See the end of the file for license conditions.

Please send Emacs bug reports to bug-gnu-emacs@gnu.org.
If possible, use M-x report-emacs-bug.

This file is about changes in Emacs version 24.

See files NEWS.23, NEWS.22, NEWS.21, NEWS.20, NEWS.19, NEWS.18,
and NEWS.1-17 for changes in older Emacs versions.

You can narrow news to a specific version by calling `view-emacs-news'
with a prefix argument or by typing C-u C-h C-n.


* Installation Changes in Emacs 24.1

** Emacs can be compiled with Gtk+ 3.0 if you pass --with-x-toolkit=gtk3
to configure.  Note that other libraries used by Emacs, RSVG and GConf,
also depend on Gtk+.  You can disable them with --without-rsvg and
--without-gconf.

** Emacs can be compiled with GnuTLS support.
This happens by default if a suitably recent version of the library is
found at build time.  To prevent this, use the configure option
`--without-gnutls'.  See below for GnuTLS features.

** Emacs can be compiled with SELinux support.
This happens by default if a suitably recent version of the library is
found at build time.  To prevent this, use the configure option
`--without-selinux'.  See below for SELinux features.

** Emacs can be compiled with ImageMagick support.
This happens by default if a suitably recent version of the library is
found at build time.  To prevent this, use the configure option
`--without-imagemagick'.  See below for ImageMagick features.
This feature is not available for the Nextstep or MS ports.

** Emacs can be compiled with libxml2 support.
This happens by default if a suitably recent version of the library is
found at build time.  To prevent this, use the configure option
`--without-xml2'.  See below for libxml2 features.

** By default, the installed Info and man pages are compressed.
You can disable this by configuring --without-compress-info.

** New configure option --with-wide-int.
With it, Emacs integers typically have 62 bits, even on 32-bit machines.
On 32-bit hosts, this raises the limit on buffer sizes from about 512 MiB
to about 2 GiB.

** New configure options: --with-mmdf, --with-mail-unlink, --with-mailhost.
These provide no new functionality, they just remove the need to edit
lib-src/Makefile by hand in order to use the associated features.

** New configure option --enable-use-lisp-union-type.
This is only useful for Emacs developers to debug certain types of bugs.
This is not a new feature; only the configure flag is new.

** The standalone programs digest-doc and sorted-doc are removed.
Emacs now uses Lisp commands `doc-file-to-man' and `doc-file-to-info'.

** The standalone program `fakemail' is removed.
If you need it, feedmail.el provides a superset of the functionality.


* Startup Changes in Emacs 24.1

** The --unibyte, --multibyte, --no-multibyte, and --no-unibyte
command line arguments, and the EMACS_UNIBYTE environment variable, no
longer have any effect.  (They were declared obsolete in Emacs 23.)

** New command line option `--no-site-lisp' removes site-lisp directories
from load-path.  -Q now implies this.  This option does not affect the
EMACSLOADPATH environment variable (and hence has no effect for
Nextstep builds).


* Changes in Emacs 24.1

** Completion

*** Many packages now use the `completion-at-point' command,
rather than implementing separate completion commands.

*** `completion-at-point' now handles tags and semantic completion.

*** Completion in a non-minibuffer now tries to detect the end of completion
and pops down the *Completions* buffer accordingly.

*** New option `completion-cycle-threshold' allows completion cycling.

*** New option `completion-category-overrides' for overriding the
default completion style in certain circumstances.

*** New completion style `substring'.

*** Completion of buffer names uses `substring' completion by default.

*** The option `widget-complete-field' has been removed.

** Mail changes

*** The first time you try sending mail, Emacs asks for a mail method.
This is implemented by a new default for `send-mail-function', which
is `sendmail-query-once'.  This offers to use the smtpmail package, or
to use the old defaults relying on external mail facilities
(`sendmail-send-it' on GNU/Linux and other Unix-like systems, and
`mailclient-send-it' on Windows).

*** Typing C-c m in the buffer made by M-x report-emacs-bug transfers
the report to your desktop's preferred mail client, if there is one.
This uses either the "xdg-email" utility, or Mac OS's "open" command.

*** See Changes in Specialized Modes and Packages for SMTPmail changes
and Mail mode changes

** Emacs server and client changes

*** New option `server-port' specifies the port for TCP Emacs servers.

*** New emacsclient argument -q/--quiet suppresses some status messages.

*** New emacsclient argument --frame-parameters specifies the frame
parameters of any newly-created graphical frame.

*** If emacsclient shuts down due to Emacs signaling an error,
its exit status is 1.

*** New emacsclient argument --parent-id ID.
This opens a client frame in parent X window ID, via XEmbed, similar
to the --parent-id argument to Emacs.

** Internationalization changes

*** Emacs now supports display and editing of bidirectional text.
Right-to-left (RTL) scripts, such as Arabic, Farsi, and Hebrew, are
displayed in the correct visual order as expected by users of those
scripts.  The display reordering is a "full bidirectionality" class
implementation of the Unicode Bidirectional Algorithm (UBA).  Buffers
with no RTL text should look exactly the same as before.

**** New buffer-local variable `bidi-display-reordering'.
To disable display reordering in a buffer, change this to nil.

**** New buffer-local variable `bidi-paragraph-direction'.
If nil (the default), Emacs determines the base direction of each
paragraph from its text, as specified by the UBA.  Setting the value

lisp/cedet/ede/emacs.el
;;; ede/emacs.el --- Special project for Emacs

;; Copyright (C) 2008-2012 Free Software Foundation, Inc.

;; Author: Eric M. Ludlam <eric@siege-engine.com>

;; This file is part of GNU Emacs.

;; GNU Emacs is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; GNU Emacs is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:
;;
;; Provide a special project type just for Emacs, cause Emacs is special.
;;
;; Identifies an Emacs project automatically.
;; Speedy ede-expand-filename based on extension.
;; Pre-populates the preprocessor map from lisp.h
;;
;; ToDo :
;; * Add "build" options.
;; * Add texinfo lookup options.
;; * Add website

(require 'ede)
(declare-function semanticdb-file-table-object "semantic/db")
(declare-function semanticdb-needs-refresh-p "semantic/db")
(declare-function semanticdb-refresh-table "semantic/db")

;;; Code:
(defvar ede-emacs-project-list nil
  "List of projects created by option `ede-emacs-project'.")

(defun ede-emacs-file-existing (dir)
  "Find a Emacs project in the list of Emacs projects.
DIR is the directory to search from."
  (let ((projs ede-emacs-project-list)
	(ans nil))
    (while (and projs (not ans))
      (let ((root (ede-project-root-directory (car projs))))
	(when (string-match (concat "^" (regexp-quote root))
			    (file-name-as-directory dir))
	  (setq ans (car projs))))
      (setq projs (cdr projs)))
    ans))

;;;###autoload
(defun ede-emacs-project-root (&optional dir)
  "Get the root directory for DIR."
  (when (not dir) (setq dir default-directory))
  (let ((case-fold-search t)
	(proj (ede-emacs-file-existing dir)))
    (if proj
	(ede-up-directory (file-name-directory
			   (oref proj :file)))
      ;; No pre-existing project.  Let's take a wild-guess if we have
      ;; an Emacs project here.
      (when (string-match "emacs[^/]*" dir)
	(let ((base (substring dir 0 (match-end 0))))
	  (when (file-exists-p (expand-file-name "src/emacs.c" base))
	      base))))))

(defun ede-emacs-version (dir)
  "Find the Emacs version for the Emacs src in DIR.
Return a tuple of ( EMACSNAME . VERSION )."
  (let ((buff (get-buffer-create " *emacs-query*"))
	(emacs "Emacs")
	(ver ""))
    (with-current-buffer buff
      (erase-buffer)
      (setq default-directory (file-name-as-directory dir))
      ;(call-process "egrep" nil buff nil "-n" "-e" "^version=" "Makefile")
      (call-process "egrep" nil buff nil "-n" "-e" "AC_INIT" "configure.in")
      (goto-char (point-min))
      ;(re-search-forward "version=\\([0-9.]+\\)")
      (cond
       ;; Maybe XEmacs?
       ((file-exists-p "version.sh")
	(setq emacs "XEmacs")
	(insert-file-contents "version.sh")
	(goto-char (point-min))
	(re-search-forward "emacs_major_version=\\([0-9]+\\)
emacs_minor_version=\\([0-9]+\\)
emacs_beta_version=\\([0-9]+\\)")
	(setq ver (concat (match-string 1) "."
			  (match-string 2) "."
			  (match-string 3)))
	)
       ;; Insert other Emacs here...

       ;; Vaguely recent version of GNU Emacs?
       (t
	(insert-file-contents "configure.in")
	(goto-char (point-min))
	(re-search-forward "AC_INIT(emacs,\\s-*\\([0-9.]+\\)\\s-*)")
	(setq ver (match-string 1))
	)
       )
      ;; Return a tuple
      (cons emacs ver))))

(defclass ede-emacs-project (ede-project eieio-instance-tracker)
  ((tracking-symbol :initform 'ede-emacs-project-list)
   )
  "Project Type for the Emacs source code."
  :method-invocation-order :depth-first)

(defun ede-emacs-load (dir &optional rootproj)
  "Return an Emacs Project object if there is a match.
Return nil if there isn't one.
Argument DIR is the directory it is created for.
ROOTPROJ is nil, since there is only one project."
  (or (ede-emacs-file-existing dir)
      ;; Doesn't already exist, so let's make one.
      (let* ((vertuple (ede-emacs-version dir)))
	(ede-emacs-project (car vertuple)
			   :name (car vertuple)
			   :version (cdr vertuple)
			   :directory (file-name-as-directory dir)
			   :file (expand-file-name "src/emacs.c"
						   dir)))
      (ede-add-project-to-global-list this)
      )
  )

;;;###autoload
(add-to-list 'ede-project-class-files
	     (ede-project-autoload "emacs"
	      :name "EMACS ROOT"
	      :file 'ede/emacs
	      :proj-file "src/emacs.c"
	      :proj-root 'ede-emacs-project-root
	      :load-type 'ede-emacs-load
	      :class-sym 'ede-emacs-project
	      :new-p nil)
	     t)

(defclass ede-emacs-target-c (ede-target)
  ()
  "EDE Emacs Project target for C code.
All directories need at least one target.")

(defclass ede-emacs-target-el (ede-target)
  ()
  "EDE Emacs Project target for Emacs Lisp code.
All directories need at least one target.")

(defclass ede-emacs-target-misc (ede-target)
  ()
  "EDE Emacs Project target for Misc files.
All directories need at least one target.")

(defmethod initialize-instance ((this ede-emacs-project)
				&rest fields)
  "Make sure the targets slot is bound."
  (call-next-method)
  (unless (slot-boundp this 'targets)
    (oset this :targets nil)))

;;; File Stuff
;;
(defmethod ede-project-root-directory ((this ede-emacs-project)
				       &optional file)
  "Return the root for THIS Emacs project with file."
  (ede-up-directory (file-name-directory (oref this file))))

(defmethod ede-project-root ((this ede-emacs-project))
  "Return my root."
  this)

(defmethod ede-find-subproject-for-directory ((proj ede-emacs-project)
					      dir)
  "Return PROJ, for handling all subdirs below DIR."
  proj)

;;; TARGET MANAGEMENT
;;
(defun ede-emacs-find-matching-target (class dir targets)
  "Find a target that is a CLASS and is in DIR in the list of TARGETS."
  (let ((match nil))
    (dolist (T targets)
      (when (and (object-of-class-p T class)
		 (string= (oref T :path) dir))
	(setq match T)
      ))
    match))

(defmethod ede-find-target ((proj ede-emacs-project) buffer)
  "Find an EDE target in PROJ for BUFFER.
If one doesn't exist, create a new one for this directory."
  (let* ((ext (file-name-extension (buffer-file-name buffer)))
	 (cls (cond ((not ext)
		     'ede-emacs-target-misc)
		    ((string-match "c\\|h" ext)
		     'ede-emacs-target-c)
		    ((string-match "elc?" ext)
		     'ede-emacs-target-el)
		    (t 'ede-emacs-target-misc)))
	 (targets (oref proj targets))
	 (dir default-directory)
	 (ans (ede-emacs-find-matching-target cls dir targets))
	 )
    (when (not ans)
      (setq ans (make-instance
		 cls
		 :name (file-name-nondirectory
			(directory-file-name dir))
		 :path dir
		 :source nil))
      (object-add-to-list proj :targets ans)
      )
    ans))

;;; UTILITIES SUPPORT.
;;
(defmethod ede-preprocessor-map ((this ede-emacs-target-c))
  "Get the pre-processor map for Emacs C code.
All files need the macros from lisp.h!"
  (require 'semantic/db)
  (let* ((proj (ede-target-parent this))
	 (root (ede-project-root proj))
	 (table (semanticdb-file-table-object
		 (ede-expand-filename root "lisp.h")))
	 (config (semanticdb-file-table-object
		  (ede-expand-filename root "config.h")))
	 filemap
	 )
    (when table
      (when (semanticdb-needs-refresh-p table)
	(semanticdb-refresh-table table))
      (setq filemap (append filemap (oref table lexical-table)))
      )
    (when config
      (when (semanticdb-needs-refresh-p config)
	(semanticdb-refresh-table config))
      (setq filemap (append filemap (oref config lexical-table)))
      )
    filemap
    ))

(defun ede-emacs-find-in-directories (name base dirs)
  "Find NAME is BASE directory sublist of DIRS."
  (let ((ans nil))
    (while (and dirs (not ans))
      (let* ((D (car dirs))
	     (ed (expand-file-name D base))
	     (ef (expand-file-name name ed)))
	(if (file-exists-p ef)
	    (setq ans ef)
	  ;; Not in this dir?  How about subdirs?
	  (let ((dirfile (directory-files ed t))
		(moredirs nil)
		)
	    ;; Get all the subdirs.
	    (dolist (DF dirfile)
	      (when (and (file-directory-p DF)
			 (not (string-match "\\.$" DF)))
		(push DF moredirs)))
	    ;; Try again.
	    (setq ans (ede-emacs-find-in-directories name ed moredirs))
	    ))
	(setq dirs (cdr dirs))))
    ans))

(defmethod ede-expand-filename-impl ((proj ede-emacs-project) name)
  "Within this project PROJ, find the file NAME.
Knows about how the Emacs source tree is organized."
  (let* ((ext (file-name-extension name))
	 (root (ede-project-root proj))
	 (dir (ede-project-root-directory root))
	 (dirs (cond
		((not ext) nil)
		((string-match "h\\|c" ext)
		 '("src" "lib-src" "lwlib"))
		((string-match "elc?" ext)
		 '("lisp"))
		((string-match "texi" ext)
		 '("doc"))
		(t nil)))
	 )
    (if (not dirs) (call-next-method)
      (ede-emacs-find-in-directories name dir dirs))
    ))

(provide 'ede/emacs)

;; Local variables:
;; generated-autoload-file: "loaddefs.el"
;; generated-autoload-load-name: "ede/emacs"
;; End:

;;; ede/emacs.el ends here
